code/modules/bitrunning/virtual_domain/modular_mob_segment.dm

```diff
@@ -0,0 +1,155 @@
+#define SPAWN_ALWAYS 100
+#define SPAWN_LIKELY 85
+#define SPAWN_UNLIKELY 35
+#define SPAWN_RARE 10
+
+/datum/modular_mob_segment
+	/// Spawn no more than this amount
+	var/max = 4
+	/// Set this to false if you want explicitly what's in the list to spawn
+	var/exact = FALSE
+	/// The list of mobs to spawn
+	var/list/mob/living/mobs = list()
+	/// The mobs spawned from this segment
+	var/list/spawned_mob_refs = list()
+	/// Chance this will spawn (1 - 100)
+	var/probability = SPAWN_LIKELY
+
+/// Spawns mobs in a circle around the location
+/datum/modular_mob_segment/proc/spawn_mobs(turf/origin)
+	if(!prob(probability))
+		return
+
+	var/total_amount = exact ? rand(1, max) : length(mobs)
+
+	shuffle_inplace(mobs)
+
+	var/turf/nearby = RANGE_TURFS(2, origin)
+	for(var/turf/tile in nearby)
+		if(tile.is_blocked_turf())
+			nearby -= tile
```

Comment:
may I suggest
```suggestion

	var/list/turf/nearby = list()
	for(var/turf/tile as anything in  RANGE_TURFS(2, origin))
		if(!tile.is_blocked_turf())
			nearby += tile
```

List subtraction is expensive, as anything because we know the type, and typing the list properly